scalanet/src/io/iohk/scalanet/peergroup/dynamictls/DynamicTLSExtension.scala

```diff
@@ -0,0 +1,255 @@
+      )
+    }
+
+    def parseAsn1EncodedValue(bytes: Array[Byte]): Attempt[SignedKey] = {
+      for {
+        signedKeyBytes <- parseAsn1EncodedBytes(bytes)
+        pub <- ExtensionPublicKey.extensionPublicKeyCodec.decodeValue(BitVector(signedKeyBytes.publicKey))
+        sig <- bits.decodeValue(BitVector(signedKeyBytes.signature))
+      } yield SignedKey(pub, sig)
+    }
+
+    /**
+      *  From libp2p tls spec:
+      *  The peer signs the concatenation of the string `libp2p-tls-handshake:` and the public key that it used to generate
+      *  the certificate carrying the libp2p Public Key Extension, using its private host key.
+      *
+      */
+    private def buildSignedKey(
+        keyType: KeyType,
+        hostKeyPair: KeyPair,
+        connectionPublicKey: BitVector,
+        secureRandom: SecureRandom
+    ): Attempt[SignedKey] = {
+      val bytesToSign = prefixAsBytes ++ connectionPublicKey
+      val signature = BitVector(CryptoUtils.signEcdsa(bytesToSign.toByteArray, hostKeyPair.getPrivate, secureRandom))
+      Attempt.fromTry(
+        ExtensionPublicKey(keyType, hostKeyPair.getPublic).map(extPublicKey => SignedKey(extPublicKey, signature))
+      )
+    }
+
+    def buildSignedKeyExtension(
+        keyType: KeyType,
+        hostKeyPair: KeyPair,
+        connectionPublicKey: BitVector,
+        secureRandom: SecureRandom
+    ): Attempt[(SignedKey, Extension)] =
+      for {
+        signedKey <- buildSignedKey(keyType, hostKeyPair, connectionPublicKey, secureRandom)
+        encoded <- toCertExtension(signedKey)
+      } yield (signedKey, encoded)
+
+    def verifySignature(signedKey: SignedKey, certPublicKey: BitVector): Boolean = {
+      val bytes = (prefixAsBytes ++ certPublicKey).toByteArray
+      CryptoUtils.verifyEcdsa(bytes, signedKey.signature.toByteArray, signedKey.publicKey.encodedPublicKey)
+    }
+
+  }
+
+  case class SignedKeyExtensionNodeData(
+      calculatedNodeId: BitVector,
+      certWithExtension: X509Certificate,
+      generatedConnectionKey: KeyPair
+  )
+
+  object SignedKeyExtensionNodeData {
+    def apply(
+        hostKeyType: KeyType,
+        hostKeyPair: KeyPair,
+        connectionKeyType: SupportedCurves,
+        secureRandom: SecureRandom
+    ): Try[SignedKeyExtensionNodeData] = {
+      // key must be one from 5.1.1. Supported Elliptic Curves Extension rfc4492, we only support subset which is also
+      // available in tls 1.3 it will ease up migration in the future
+      val connectionKeyPair = CryptoUtils.genTlsSupportedKeyPair(secureRandom, connectionKeyType)
+
+      // safe to call get, as key is generated by us
+      val connectionKeyPairPublicKeyAsBytes = CryptoUtils.getEcPublicKey(connectionKeyPair.getPublic).get
+
+      // Certificate will be valid for next 100 years, the same value is used in libp2p go implementation
+      val today = LocalDateTime.now(Clock.systemUTC())
+      val beforeDate = Date.from(today.minusMonths(1).toInstant(ZoneOffset.UTC))
```

Comment:
is `libp2p go ` uses also one month before date?

Reply:
nope, i used it to avoid situation that validation fails becouse someones system clock is skewed.

Reply:
i will change it to day before date, it should be enough for any weirdness in times.